src/sympc/protocol/falcon/falcon.py

```diff
@@ -73,93 +77,293 @@ def __eq__(self, other: Any) -> bool:
-        Falcon, requires parties to be able to communication between each other.
-
         Args:
             x (MPCTensor): Secret
             y (MPCTensor): Another secret
             session (Session): Session the tensors belong to
+            op_str (str): Operation string.
+            kwargs_ (dict): Kwargs for some operations like conv2d
 
         Returns:
-            shares (ReplicatedSharedTensor): Shares in terms of ReplicatedSharedTensor.
+            result(MPCTensor): Result of the operation.
 
         Raises:
             ValueError: Raised when number of parties are not three.
-            NotImplementedError: Raised when implementation not present
-
+            ValueError : Raised when invalid security_type is provided.
         """
         if len(session.parties) != 3:
             raise ValueError("Falcon requires 3 parties")
 
         result = None
 
         if session.protocol.security_type == "semi-honest":
-            args = []
-            for index in range(0, 3):
-                args.append(
-                    [
-                        x.share_ptrs[index],
-                        y.share_ptrs[index],
-                    ]
-                )
-            z_shares_ptrs = parallel_execution(
-                Falcon.compute_zvalue_and_add_mask, session.parties
-            )(args)
-
-            result = MPCTensor(shares=z_shares_ptrs, session=x.session)
-            result.shape = MPCTensor._get_shape("mul", x.shape, y.shape)  # for prrs
-            result = ABY3.truncate(result, session)
-
+            result = Falcon.mul_semi_honest(x, y, session, op_str, kwargs_)
+        elif session.protocol.security_type == "malicious":
+            result = Falcon.mul_malicious(x, y, session, op_str, kwargs_)
         else:
-            raise NotImplementedError(
-                f"mult operation not implemented for {session.protocol.security_type} setting"
+            raise ValueError("Invalid security_type for Falcon multiplication")
+
+        result = ABY3.truncate(result, session)
+
+        return result
+
+    @staticmethod
+    def mul_semi_honest(
+        x: MPCTensor,
+        y: MPCTensor,
+        session: Session,
+        op_str: str,
+        kwargs_: Dict[Any, Any],
+        truncate: bool = True,
+    ) -> MPCTensor:
+        """Falcon semihonest multiplication.
+
+        Performs Falcon's mul implementation, add masks and performs resharing.
+
+        Args:
+            x (MPCTensor): Secret
+            y (MPCTensor): Another secret
+            session (Session): Session the tensors belong to
+            op_str (str): Operation string.
+            kwargs_ (dict): Kwargs for some operations like conv2d
+            truncate (bool) : Applies truncation on the result if set.
+
+        Returns:
+            MPCTensor: Result of the operation.
+        """
+        result = None
+        args = []
+        for index in range(0, 3):
```

Comment:
NIT:
```
args = [(x_share, y_xshare, op_str) for x_share in y_share in zip(x.share_ptrs, y.share_ptrs)]
```

Reply:
@kamathhrishi  , had this exact change in his refactoring PR, https://github.com/OpenMined/SyMPC/pull/255 , so I thought not to do here,Shall I change here or we will address in the refactor PR.

Reply:
I have addressed it so it will change after I merge, so you can keep it like that.

Reply:
I have modified it after the merge.